---a/net/minecraft/inventory/SlotFurnaceOutput.java
+++b/net/minecraft/inventory/SlotFurnaceOutput.java
@@ -6,73 +6,66 @@
 import net.minecraft.item.ItemStack;
 import net.minecraft.item.crafting.FurnaceRecipes;
 import net.minecraft.stats.AchievementList;
+import net.minecraft.tileentity.TileEntityFurnace;
 import net.minecraft.util.MathHelper;
 
-public class SlotFurnaceOutput extends Slot
-{
+import org.bukkit.entity.Player;
+import org.bukkit.event.inventory.FurnaceExtractEvent;
+
+public class SlotFurnaceOutput extends Slot {
+
     private EntityPlayer thePlayer;
     private int field_75228_b;
     private static final String __OBFID = "CL_00002183";
 
-    public SlotFurnaceOutput(EntityPlayer player, IInventory p_i45793_2_, int slotIndex, int xPosition, int yPosition)
-    {
+    public SlotFurnaceOutput(EntityPlayer player, IInventory p_i45793_2_, int slotIndex, int xPosition, int yPosition) {
         super(p_i45793_2_, slotIndex, xPosition, yPosition);
         this.thePlayer = player;
     }
 
-    public boolean isItemValid(ItemStack stack)
-    {
+    public boolean isItemValid(ItemStack stack) {
         return false;
     }
 
-    public ItemStack decrStackSize(int amount)
-    {
-        if (this.getHasStack())
-        {
+    public ItemStack decrStackSize(int amount) {
+        if (this.getHasStack()) {
             this.field_75228_b += Math.min(amount, this.getStack().stackSize);
         }
 
         return super.decrStackSize(amount);
     }
 
-    public void onPickupFromSlot(EntityPlayer playerIn, ItemStack stack)
-    {
+    public void onPickupFromSlot(EntityPlayer playerIn, ItemStack stack) {
         this.onCrafting(stack);
         super.onPickupFromSlot(playerIn, stack);
     }
 
-    protected void onCrafting(ItemStack stack, int amount)
-    {
+    protected void onCrafting(ItemStack stack, int amount) {
         this.field_75228_b += amount;
         this.onCrafting(stack);
     }
 
-    protected void onCrafting(ItemStack stack)
-    {
+    protected void onCrafting(ItemStack stack) {
         stack.onCrafting(this.thePlayer.worldObj, this.thePlayer, this.field_75228_b);
 
-        if (!this.thePlayer.worldObj.isRemote)
-        {
+        if (!this.thePlayer.worldObj.isRemote) {
             int i = this.field_75228_b;
             float f = FurnaceRecipes.instance().getSmeltingExperience(stack);
             int j;
 
-            if (f == 0.0F)
-            {
+            if (f == 0.0F) {
                 i = 0;
-            }
-            else if (f < 1.0F)
-            {
-                j = MathHelper.floor_float((float)i * f);
+            } else if (f < 1.0F) {
+                j = MathHelper.floor_float((float) i * f);
 
-                if (j < MathHelper.ceiling_float_int((float)i * f) && Math.random() < (double)((float)i * f - (float)j))
-                {
+                if (j < MathHelper.ceiling_float_int((float) i * f) && Math.random() < (double) ((float) i * f - (float) j)) {
                     ++j;
                 }
 
                 i = j;
             }
 
-            while (i > 0)
-            {
+            // CraftBukkit start - fire FurnaceExtractEvent
+            Player player = (Player) thePlayer.getBukkitEntity();
+            TileEntityFurnace furnace = ((TileEntityFurnace) this.inventory);
+            org.bukkit.block.Block block = thePlayer.worldObj.getWorld().getBlockAt(furnace.pos.getX(), furnace.pos.getY(), furnace.pos.getZ());
+
+            FurnaceExtractEvent event = new FurnaceExtractEvent(player, block, org.bukkit.craftbukkit.util.CraftMagicNumbers.getMaterial(stack.getItem()), stack.stackSize, i);
+            thePlayer.worldObj.getServer().getPluginManager().callEvent(event);
+
+            i = event.getExpToDrop();
+            // CraftBukkit end
+
+            while (i > 0) {
                 j = EntityXPOrb.getXPSplit(i);
                 i -= j;
                 this.thePlayer.worldObj.spawnEntityInWorld(new EntityXPOrb(this.thePlayer.worldObj, this.thePlayer.posX, this.thePlayer.posY + 0.5D, this.thePlayer.posZ + 0.5D, j));
@@ -83,15 +83,12 @@
 
         this.field_75228_b = 0;
 
-        if (stack.getItem() == Items.iron_ingot)
-        {
+        if (stack.getItem() == Items.iron_ingot) {
             this.thePlayer.triggerAchievement(AchievementList.acquireIron);
         }
 
-        if (stack.getItem() == Items.cooked_fish)
-        {
+        if (stack.getItem() == Items.cooked_fish) {
             this.thePlayer.triggerAchievement(AchievementList.cookFish);
         }
     }
-}
+}
